File: packages/shared/src/pricing/reward.ts
// The reward function — ReLoop's thesis as math, mirrored 1:1 by reward.py so the
// TypeScript runtime and the Python trainer optimise the SAME objective. This is the
// quantity the XGBoost reward model learns to predict (E[reward | features, arm]).
//
//   sold     →  margin − holding·days + carbon-credit-if-local
//   rerouted →  −penalty (warehouse worse than recycle: the item moved + still no sale)
//   listed   →  0 (intermediate; partial-credit signal during training, not terminal)

import type { PricingOutcome } from './types.js';

/** Reward weights — tune per business decision (kept in sync with reward.py). */
export type RewardConfig = {
  holdingCostPerDay: number; // ₹ cost of capital per day the item sits unsold
  handlingCost: number; // ₹ fixed cost per transaction
  carbonCreditLocal: number; // ₹ equivalent credit for a local sale (no warehouse leg)
  warehousePenalty: number; // ₹ penalty for falling through to warehouse
  recyclePenalty: number; // ₹ penalty for falling through to recycle/donate
};

export const DEFAULT_REWARD_CONFIG: RewardConfig = {
  holdingCostPerDay: 8,
  handlingCost: 120,
  carbonCreditLocal: 45,
  warehousePenalty: 200,
  recyclePenalty: 100,
};

export function computeReward(
  outcome: PricingOutcome,
  config: RewardConfig = DEFAULT_REWARD_CONFIG,
): number {
  if (outcome.sold) {
    const margin = outcome.finalPrice - config.handlingCost;
    const holdingPenalty = config.holdingCostPerDay * outcome.daysOnMarket;
    const carbonBonus = outcome.soldLocally ? config.carbonCreditLocal : 0;
    return margin - holdingPenalty + carbonBonus;
  }

  if (outcome.rerouted) {
    // Fell through to donate / recycle / warehouse.
    const penalty =
      outcome.rerouteDestination === 'warehouse'
        ? config.warehousePenalty
        : config.recyclePenalty;
    return -penalty;
  }

  // Still listed — intermediate signal (used for partial credit during training,
  // never as a terminal reward).
  return 0;
}

File: ml/pricing/reloop_pricing/pricing/reward.py
"""Reward function — ReLoop's thesis as math. 1:1 mirror of
packages/shared/src/pricing/reward.ts. The TS runtime and this trainer optimise the
SAME objective; this is the quantity XGBoost learns to predict per price arm.

  sold     →  margin − holding·days + carbon-credit-if-local
  rerouted →  −penalty (warehouse worse than recycle)
  listed   →  0 (intermediate; partial credit only, never terminal)
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Mapping


@dataclass(frozen=True)
class RewardConfig:
    holding_cost_per_day: float = 8.0
    handling_cost: float = 120.0
    carbon_credit_local: float = 45.0
    warehouse_penalty: float = 200.0
    recycle_penalty: float = 100.0


DEFAULT_REWARD_CONFIG = RewardConfig()


def compute_reward(outcome: Mapping, config: RewardConfig = DEFAULT_REWARD_CONFIG) -> float:
    """outcome keys: sold, finalPrice, daysOnMarket, soldLocally, rerouted, rerouteDestination."""
    if outcome.get("sold"):
        margin = float(outcome["finalPrice"]) - config.handling_cost
        holding_penalty = config.holding_cost_per_day * float(outcome.get("daysOnMarket", 0))
        carbon_bonus = config.carbon_credit_local if outcome.get("soldLocally") else 0.0
        return margin - holding_penalty + carbon_bonus

    if outcome.get("rerouted"):
        penalty = (
            config.warehouse_penalty
            if outcome.get("rerouteDestination") == "warehouse"
            else config.recycle_penalty
        )
        return -penalty

    return 0.0
